File: packages/skillet-eval-python/src/skill_eval/config/paths.py
from __future__ import annotations

from pathlib import Path

PROJECT_ROOT_MARKERS = [".git", "package.json", "pyproject.toml", "Cargo.toml"]


def find_project_root(start_dir: str) -> str:
    dir_path = Path(start_dir).resolve()
    while True:
        for marker in PROJECT_ROOT_MARKERS:
            if (dir_path / marker).exists():
                return str(dir_path)
        parent = dir_path.parent
        if parent == dir_path:
            return str(Path(start_dir).resolve())
        dir_path = parent


def resolve_skill_paths(skill_path: str, evals_override: str | None = None) -> dict[str, str]:
    abs_path = Path(skill_path).resolve()
    skill_name = abs_path.name
    project_root = find_project_root(str(abs_path))
    evals_file = (
        str(Path(evals_override).resolve())
        if evals_override
        else str(abs_path / "evals.json")
    )
    return {
        "skill_dir": str(abs_path),
        "skill_name": skill_name,
        "skill_file": str(abs_path / "SKILL.md"),
        "evals_file": evals_file,
        "results_dir": str(Path(project_root) / ".skillet-evals" / "results" / skill_name),
    }

File: packages/skillet/src/eval/config/paths.ts
import fs from "node:fs";
import path from "node:path";

const PROJECT_ROOT_MARKERS = [".git", "package.json", "pyproject.toml", "Cargo.toml"];

/** Find the project root by walking up from startDir looking for markers. */
export function findProjectRoot(startDir: string): string {
	let dir = path.resolve(startDir);
	while (true) {
		for (const marker of PROJECT_ROOT_MARKERS) {
			if (fs.existsSync(path.join(dir, marker))) return dir;
		}
		const parent = path.dirname(dir);
		if (parent === dir) return path.resolve(startDir);
		dir = parent;
	}
}

/**
 * Pick the canonical evals.json location for a skill directory.
 *
 * Skillet supports two layouts:
 * - Flat (default for single-skill repos): `<skill>/evals.json` with fixtures
 *   colocated under `<skill>/fixtures/`.
 * - Nested (skill-creator convention): `<skill>/evals/evals.json` with
 *   fixtures under `<skill>/evals/files/`.
 *
 * The flat layout takes precedence when both exist; the nested layout is used
 * as a fallback when only it exists.
 */
function resolveEvalsFile(skillDir: string): string {
	const flat = path.join(skillDir, "evals.json");
	if (fs.existsSync(flat)) return flat;
	const nested = path.join(skillDir, "evals", "evals.json");
	if (fs.existsSync(nested)) return nested;
	return flat;
}

/** Resolve paths within a skill directory by convention. */
export function resolveSkillPaths(skillPath: string, evalsOverride?: string) {
	const abs = path.resolve(skillPath);
	const skillName = path.basename(abs);
	const projectRoot = findProjectRoot(abs);
	const evalsFile = evalsOverride ? path.resolve(evalsOverride) : resolveEvalsFile(abs);
	return {
		skillDir: abs,
		skillName,
		skillFile: path.join(abs, "SKILL.md"),
		evalsFile,
		evalsDir: path.dirname(evalsFile),
		resultsDir: path.join(projectRoot, ".skillet-evals", "results", skillName),
	};
}
